File: entropy_engine.py
"""动态熵减决策引擎（纯本地 Python 实现）。

数据来源：
- algorithm_prior.json：初始先验 + 可调 params
- feature_algorithm_matrix.json：29 个特征问题（20 基础 + 9 教师共识）+ 120 算法 profile + 四方向权重

只使用 math.log2、加减乘除、归一化，不调用任何模型或外部 API。

性能设计：
- 构造函数把 profile / direction_weights 预计算成矩阵，避免每次更新都反复查 dict。
- feature_by_id 使用索引字典。
"""

import json
import math
import os

from resource_paths import find_data

REPO_ROOT = os.path.dirname(os.path.abspath(__file__))
DEFAULT_PRIOR = find_data("algorithm_prior.json")
DEFAULT_MATRIX = find_data("feature_algorithm_matrix.json")
DEFAULT_HEURISTICS = find_data("heuristics.json")


def _norm(weights):
    total = sum(weights)
    if total <= 0:
        n = len(weights)
        return [1.0 / n] * n if n else []
    return [w / total for w in weights]


def _entropy(weights):
    h = 0.0
    for w in weights:
        if w > 0:
            h -= w * math.log2(w)
    return h


class EntropyEngine:
    def __init__(self, prior_path=None, matrix_path=None, heuristics_path=None):
        prior_path = prior_path or DEFAULT_PRIOR
        matrix_path = matrix_path or DEFAULT_MATRIX
        heuristics_path = heuristics_path or DEFAULT_HEURISTICS
        with open(prior_path, encoding="utf-8") as f:
            self.prior = json.load(f)
        with open(matrix_path, encoding="utf-8") as f:
            self.matrix = json.load(f)
        with open(heuristics_path, encoding="utf-8") as f:
            self.heuristics = json.load(f)

        self.params = self.prior.get("params", {})
        self.features = self.matrix.get("features", [])
        self.directions = self.matrix.get("directions", [])
        self.algorithms = self.matrix.get("algorithms", [])

        priors = {a["algorithm_name"]: a.get("prior_probability", 0) for a in self.prior.get("algorithms", [])}
        for alg in self.algorithms:
            alg.setdefault("prior_probability", priors.get(alg["algorithm_name"], 0))

        # ---- 预计算矩阵：让每次更新只做乘法，不做 dict 查找 ----
        self._feature_index = {f["id"]: i for i, f in enumerate(self.features)}
        self._feature_ids = [f["id"] for f in self.features]
        # 列优先存储：每个特征一列 P(feature | algorithm)，
        # 避免遍历“算法 × 特征”时反复跨行访问。
        self._profile_columns = [
            [float(alg.get("profile", {}).get(fid, 0.5)) for alg in self.algorithms]
            for fid in self._feature_ids
        ]
        self._direction_columns = {
            d: [float(alg.get("direction_weights", {}).get(d, 0.0)) for alg in self.algorithms]
            for d in self.directions
        }

    # ---------- 基础工具 ----------
    def initial_weights(self):
        return _norm([max(0.0, a.get("prior_probability", 0)) for a in self.algorithms])

    def feature_by_id(self, fid):
        idx = self._feature_index.get(fid)
        if idx is None:
            return None
        return self.features[idx]

    def feature_index(self, fid):
        return self._feature_index.get(fid)

    def profile(self, alg_index, fid):
        """保留原接口：按特征 id 取 P(feature | algorithm)。"""
        idx = self._feature_index.get(fid)
        if idx is None:
            return 0.5
        return self._profile_columns[idx][alg_index]

    def entropy(self, weights):
        return _entropy(_norm(weights))

    def normalize(self, weights):
        return _norm(weights)

    def _py(self, fid_idx, w):
        """P(yes | feature) 对已归一化权重求和。"""
        column = self._profile_columns[fid_idx]
        total = 0.0
        for i, wi in enumerate(w):
            total += wi * column[i]
        return total

    def answer_probability(self, weights, fid, answer):
        if answer == "uncertain":
            # uncertain 的“惊讶度”按参数化弱证据混合期望
            decay = self.params.get("uncertain_decay", 0.5)
            py = self.answer_probability(weights, fid, "yes")
            pn = self.answer_probability(weights, fid, "no")
            return decay * py + (1.0 - decay) * pn
        w = self.normalize(weights)
        fid_idx = self._feature_index.get(fid)
        if fid_idx is None:
            return 1.0
        if answer == "yes":
            return self._py(fid_idx, w)
        return 1.0 - self._py(fid_idx, w)

    def _posterior_from_normalized(self, w, fid_idx, answer):
        if answer == "uncertain":
            decay = self.params.get("uncertain_decay", 0.5)
            wy = self._posterior_from_normalized(w, fid_idx, "yes")
            wn = self._posterior_from_normalized(w, fid_idx, "no")
            return self.normalize([decay * a + (1.0 - decay) * b for a, b in zip(wy, wn, strict=False)])
        column = self._profile_columns[fid_idx]
        out = []
        if answer == "yes":
            for i, wi in enumerate(w):
                out.append(wi * column[i])
        else:
            for i, wi in enumerate(w):
                out.append(wi * (1.0 - column[i]))
        return self.normalize(out)

    def posterior(self, weights, fid, answer):
        fid_idx = self._feature_index.get(fid)
        if fid_idx is None:
            return self.normalize(weights)
        w = self.normalize(weights)
        return self._posterior_from_normalized(w, fid_idx, answer)

    def information_gain(self, weights, fid):
        fid_idx = self._feature_index.get(fid)
        if fid_idx is None:
            return 0.0
        w = self.normalize(weights)
        py = self._py(fid_idx, w)
        pn = 1.0 - py
        if py <= 0 or pn <= 0:
            return 0.0
        hy = _entropy(self._posterior_from_normalized(w, fid_idx, "yes"))
        hn = _entropy(self._posterior_from_normalized(w, fid_idx, "no"))
        expected = py * hy + pn * hn
        return _entropy(w) - expected

    def choose_next(self, weights, asked):
        best_id = None
        best_ig = -1.0
        w = self.normalize(weights)
        w_entropy = _entropy(w)
        for fid_idx, fid in enumerate(self._feature_ids):
            if fid in asked:
                continue
            py = self._py(fid_idx, w)
            pn = 1.0 - py
            if py <= 0 or pn <= 0:
                continue
            hy = _entropy(self._posterior_from_normalized(w, fid_idx, "yes"))
            hn = _entropy(self._posterior_from_normalized(w, fid_idx, "no"))
            ig = w_entropy - (py * hy + pn * hn)
            if ig > best_ig:
                best_ig = ig
                best_id = fid
        return best_id, best_ig

    def should_stop(self, weights, asked):
        h = self.entropy(weights)
        if h < self.params.get("entropy_stop_threshold", 0.45):
            return True, "entropy"
        if len(asked) >= self.params.get("max_questions", 12):
            return True, "max_questions"
        best_id, best_ig = self.choose_next(weights, asked)
        if best_id is None:
            return True, "exhausted"
        if best_ig < self.params.get("ig_stop_threshold", 0.03):
            return True, "ig"
        return False, best_id

    def answer(self, weights, asked, fid, answer):
        new_weights = self.posterior(weights, fid, answer)
        surprise = self.answer_probability(weights, fid, answer)
        new_asked = list(asked) + [fid]
        return new_weights, new_asked, surprise

    def heuristic_direction(self, name):
        for h in self.heuristics.get("directions", []):
            if h.get("id") == name:
                return h
        return None

    def top_algorithms(self, weights, threshold=None, limit=None):
        """返回后验权重 >= threshold 的候选算法（按权重降序），threshold 默认取 params。"""
        if threshold is None:
            threshold = self.params.get("algorithm_weight_threshold", 0.02)
        if limit is None:
            limit = self.params.get("max_algorithm_list", 12)
        w = self.normalize(weights)
        rows = []
        for i, wi in enumerate(w):
            if wi >= threshold:
                rows.append({
                    "algorithm_name": self.algorithms[i]["algorithm_name"],
                    "weight": wi,
                })
        rows.sort(key=lambda r: r["weight"], reverse=True)
        return rows[:limit]

    def realtime_top(self, weights, n=None):
        """问题流实时展示用的 Top N 候选算法。"""
        if n is None:
            n = self.params.get("realtime_top_n", 3)
        w = self.normalize(weights)
        rows = sorted(
            ({"algorithm_name": self.algorithms[i]["algorithm_name"], "weight": wi}
             for i, wi in enumerate(w)),
            key=lambda r: r["weight"], reverse=True,
        )
        return rows[:n]

    def direction_probs(self, weights):
        w = self.normalize(weights)
        out = {}
        for d, column in self._direction_columns.items():
            total = 0.0
            for i, wi in enumerate(w):
                total += wi * column[i]
            out[d] = total
        # 归一化，方便展示
        s = sum(out.values()) or 1.0
        return {d: v / s for d, v in out.items()}

    def top_directions(self, weights, top=4):
        probs = self.direction_probs(weights)
        return sorted(probs.items(), key=lambda kv: kv[1], reverse=True)[:top]

File: mobile/www/entropy_engine.js
/* 动态熵减决策引擎（纯本地 JS）。
 * 输入 window.ENTROPY_DATA，不调用任何网络/模型。
 */
(function (global) {
  function normalize(weights) {
    var total = 0;
    for (var i = 0; i < weights.length; i++) total += weights[i];
    if (total <= 0) {
      var out = new Array(weights.length);
      for (var j = 0; j < weights.length; j++) out[j] = 1 / weights.length;
      return out;
    }
    // 与 Python _norm 对齐：返回新数组，绝不原地修改调用方传入的权重。
    var result = new Array(weights.length);
    for (var k = 0; k < weights.length; k++) result[k] = weights[k] / total;
    return result;
  }

  function entropy(weights) {
    // 与 Python EntropyEngine.entropy 对齐：先归一化再计算熵。
    var normalized = normalize(weights);
    var h = 0;
    for (var i = 0; i < normalized.length; i++) {
      var w = normalized[i];
      if (w > 0) h -= w * Math.log2(w);
    }
    return h;
  }

  function makeEngine(data) {
    var features = data.features || [];
    var algorithms = data.algorithms || [];
    var params = data.params || {};
    var directions = data.directions || [];

    var featureIds = features.map(function (f) { return f.id; });
    var featureIndex = {};
    for (var fi = 0; fi < featureIds.length; fi++) featureIndex[featureIds[fi]] = fi;

    // 列优先预计算：避免每次回答都在 120×20 个对象上做属性查找。
    var profileColumns = [];
    for (var fIdx = 0; fIdx < featureIds.length; fIdx++) {
      var fid = featureIds[fIdx];
      var col = [];
      for (var aIdx = 0; aIdx < algorithms.length; aIdx++) {
        var p = (algorithms[aIdx].profile || {})[fid];
        col.push(typeof p === "number" ? p : 0.5);
      }
      profileColumns.push(col);
    }

    var directionColumns = {};
    for (var dIdx = 0; dIdx < directions.length; dIdx++) {
      var dName = directions[dIdx];
      var dCol = [];
      for (var a2 = 0; a2 < algorithms.length; a2++) {
        dCol.push((algorithms[a2].direction_weights || {})[dName] || 0);
      }
      directionColumns[dName] = dCol;
    }

    function profile(algIndex, fid) {
      var idx = featureIndex[fid];
      if (idx == null) return 0.5;
      return profileColumns[idx][algIndex];
    }

    function answerProbabilityNorm(w, fidIdx, answer) {
      if (answer === "uncertain") {
        var decay = params.uncertain_decay == null ? 0.5 : params.uncertain_decay;
        return decay * answerProbabilityNorm(w, fidIdx, "yes")
          + (1 - decay) * answerProbabilityNorm(w, fidIdx, "no");
      }
      var s = 0;
      var col = profileColumns[fidIdx];
      if (answer === "yes") {
        for (var i = 0; i < w.length; i++) s += w[i] * col[i];
      } else {
        for (var j = 0; j < w.length; j++) s += w[j] * (1 - col[j]);
      }
      return s;
    }

    function posteriorNorm(w, fidIdx, answer) {
      if (answer === "uncertain") {
        var decay = params.uncertain_decay == null ? 0.5 : params.uncertain_decay;
        var wy = posteriorNorm(w, fidIdx, "yes");
        var wn = posteriorNorm(w, fidIdx, "no");
        var mix = [];
        for (var m = 0; m < wy.length; m++) mix.push(decay * wy[m] + (1 - decay) * wn[m]);
        return normalize(mix);
      }
      var col = profileColumns[fidIdx];
      var out = [];
      for (var i = 0; i < w.length; i++) {
        var p = answer === "yes" ? col[i] : (1 - col[i]);
        out.push(w[i] * p);
      }
      return normalize(out);
    }

    function answerProbability(weights, fid, answer) {
      var idx = featureIndex[fid];
      if (idx == null) return 1;
      return answerProbabilityNorm(normalize(weights.slice()), idx, answer);
    }

    function posterior(weights, fid, answer) {
      var idx = featureIndex[fid];
      if (idx == null) return normalize(weights.slice());
      return posteriorNorm(normalize(weights.slice()), idx, answer);
    }

    function informationGain(weights, fid) {
      var idx = featureIndex[fid];
      if (idx == null) return 0;
      var w = normalize(weights.slice());
      var py = answerProbabilityNorm(w, idx, "yes");
      var pn = 1 - py;
      if (py <= 0 || pn <= 0) return 0;
      var hy = entropy(posteriorNorm(w, idx, "yes"));
      var hn = entropy(posteriorNorm(w, idx, "no"));
      return entropy(w) - (py * hy + pn * hn);
    }

    function chooseNext(weights, asked) {
      var bestId = null;
      var bestIg = -1;
      var w = normalize(weights.slice());
      var wEntropy = entropy(w);
      for (var i = 0; i < featureIds.length; i++) {
        var fid = featureIds[i];
        if (asked.indexOf(fid) >= 0) continue;
        var py = answerProbabilityNorm(w, i, "yes");
        var pn = 1 - py;
        if (py <= 0 || pn <= 0) continue;
        var hy = entropy(posteriorNorm(w, i, "yes"));
        var hn = entropy(posteriorNorm(w, i, "no"));
        var ig = wEntropy - (py * hy + pn * hn);
        if (ig > bestIg) {
          bestIg = ig;
          bestId = fid;
        }
      }
      return { id: bestId, ig: bestIg };
    }

    function shouldStop(weights, asked) {
      var h = entropy(normalize(weights.slice()));
      var paramsStop = params.entropy_stop_threshold == null ? 0.45 : params.entropy_stop_threshold;
      if (h < paramsStop) return { stop: true, reason: "entropy" };
      var maxQ = params.max_questions == null ? 12 : params.max_questions;
      if (asked.length >= maxQ) return { stop: true, reason: "max" };
      var next = chooseNext(weights, asked);
      if (!next.id) return { stop: true, reason: "exhausted" };
      var igStop = params.ig_stop_threshold == null ? 0.03 : params.ig_stop_threshold;
      if (next.ig < igStop) return { stop: true, reason: "ig" };
      return { stop: false, reason: next.id, ig: next.ig };
    }

    function directionProbs(weights) {
      var w = normalize(weights.slice());
      var out = {};
      var total = 0;
      for (var d = 0; d < directions.length; d++) {
        var name = directions[d];
        var column = directionColumns[name];
        var s = 0;
        for (var i = 0; i < w.length; i++) s += w[i] * column[i];
        out[name] = s;
        total += s;
      }
      if (total <= 0) total = 1;
      for (var key in out) out[key] = out[key] / total;
      return out;
    }

    function topAlgorithms(weights, threshold, limit) {
      if (threshold == null) threshold = params.algorithm_weight_threshold == null ? 0.02 : params.algorithm_weight_threshold;
      if (limit == null) limit = params.max_algorithm_list == null ? 12 : params.max_algorithm_list;
      var w = normalize(weights.slice());
      var rows = [];
      for (var i = 0; i < w.length; i++) {
        if (w[i] >= threshold) {
          rows.push({ algorithm_name: algorithms[i].algorithm_name, weight: w[i] });
        }
      }
      rows.sort(function (a, b) { return b.weight - a.weight; });
      return rows.slice(0, limit);
    }

    function realtimeTop(weights, n) {
      if (n == null) n = params.realtime_top_n == null ? 3 : params.realtime_top_n;
      var w = normalize(weights.slice());
      var rows = [];
      for (var i = 0; i < w.length; i++) rows.push({ algorithm_name: algorithms[i].algorithm_name, weight: w[i] });
      rows.sort(function (a, b) { return b.weight - a.weight; });
      return rows.slice(0, n);
    }

    return {
      features: features,
      algorithms: algorithms,
      params: params,
      normalize: normalize,
      entropy: entropy,
      answerProbability: answerProbability,
      posterior: posterior,
      informationGain: informationGain,
      chooseNext: chooseNext,
      shouldStop: shouldStop,
      directionProbs: directionProbs,
      topAlgorithms: topAlgorithms,
      realtimeTop: realtimeTop,
      initialWeights: function () {
        // 与 Python initial_weights 对齐：先取非负先验，再做归一化。
        return normalize(algorithms.map(function (a) {
          return Math.max(0, a.prior_probability || 0);
        }));
      }
    };
  }

  global.EntropyEngine = { create: makeEngine };
})(window);
